Added network module

ar4s opened the pull request:
Network module contains models for network and IP and basic IP Address
Management.

mkurek commented on src/ralph/data_center/admin.py:
```diff
@@ -37,12 +45,57 @@
 from ralph.supports.models import BaseObjectsSupport
 
 
+class ParentChangeMixin(MemorizeBeforeStateMixin):
```
![](http://www.quickmeme.com/img/d3/d36ad81bae0a8d69cd91f81065d8b0a19d9dfc1b63753093992282d6aa8b35dd.jpg)

ar4s replied:
done

xliiv commented:
:ok:

xliiv commented on src/ralph/data_center/models/choices.py:
```diff
@@ -54,3 +54,10 @@ class ConnectionType(Choices):
+    _ = Choices.Choice
+
+    used = _('DHCP (used)')
+    reserved = _('reserved').extra(help_text=_('Exclude from DHCP'))
```
No need of `free` status or something?

ar4s replied:
Nope, free == not in DB

xliiv replied:
:ok:

xliiv commented on src/ralph/lib/polymorphic/models.py:
```diff
@@ -180,3 +180,7 @@ def save(self, *args, **kwargs):
         super().save(*args, **kwargs)
+
+    @property
+    def last_descendant(self):
```
I don't see use of this.

ar4s replied:
I know but we probably use this in the future.

xliiv replied:
:ok:

xliiv commented on src/ralph/lib/mixins/admin.py:
```diff
@@ -0,0 +1,9 @@
+
+    def save_model(self, request, obj, form, change):
+        """
+        Given a model instance save it to the database.
```
This docstring is too smart too me ;)

ar4s replied:
https://github.com/django/django/blob/master/django/contrib/admin/options.py#L999-L1003

I will describe class and remove this irrelevant docstring.

szok commented on src/ralph/data_center/admin.py:
```diff
@@ -316,12 +366,113 @@ class DiscoveryQueueAdmin(RalphAdmin):
+    add_form = AddNetworkForm
+
+    add_message = _('Network added to <a href="{}" _target="blank">{}</a>')
+    change_message = _('Network reassigned from network <a href="{}" target="_blank">{}</a> to <a href="{}" target="_blank">{}</a>')  # noqa
```
this format is ok for django translate?

ar4s replied:
yep

szok commented on src/ralph/data_center/admin.py:
```diff
@@ -316,12 +366,113 @@ class DiscoveryQueueAdmin(RalphAdmin):
+        """
+        bottom_margin = form.cleaned_data.get('bottom_margin', None)
+        top_margin = form.cleaned_data.get('top_margin', None)
+        if all([bottom_margin, top_margin]):
```
Why this is better than:

```
if bottom_margin and top_margin:
```

?

szok commented on src/ralph/data_center/tests/test_models.py:
```diff
@@ -19,137 +21,210 @@
         )
         self.net5 = Network.objects.create(
-            name="test5",
-            address="192.169.133.0/24",
```
:+1:

szok commented on src/ralph/data_importer/resources.py:
```diff
@@ -576,8 +577,8 @@ class OperationResource(RalphModelResource):
         column_name='base_objects',
-        attribute='base_objects',
-        widget=widgets.ManyToManyWidget(base.BaseObject),
+        attribute='asset',
```
Why asset? 
Operation model has attribute: base_objects.

ar4s replied:
My mistake :)

xliiv commented on src/ralph/data_center/tests/test_models.py:
```diff
@@ -19,137 +21,210 @@
+            name='ip_address_should_return_network',
+            address=network,
+        )
+        assigned_network = None
```
Useless?

mkurek commented on src/ralph/lib/mixins/admin.py:
```diff
@@ -0,0 +1,59 @@
+    Mixin display message when parent was changed. Works with admin when model
+    has specified `_parent_attr` attribute.
+    """
+    add_message = None
```
maybe '' would be better here?

mkurek commented on src/ralph/lib/mixins/admin.py:
```diff
@@ -0,0 +1,59 @@
+        message = None
+        old_parent = getattr(obj._before_state, self.parent_attr)
+        if parent and old_parent != parent:
+            old_url = old_parent and old_parent.get_absolute_url() or '#'
```
use parentheses here; btw why '#'?

mkurek commented on src/ralph/networks/__init__.py:
```diff
@@ -0,0 +1,17 @@
+# from ralph.networks.models.networks import (
```
what about it?

mkurek commented on src/ralph/networks/admin.py:
```diff
@@ -0,0 +1,147 @@
+    pass
+
+
+# class AddNetworkForm(RalphAdminFormMixin, ModelForm):
```
?

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+        verbose_name_plural = _('network kinds')
+        ordering = ('name',)
+
+    def __str__(self):
```
isn't it default behaviour of `NamedMixin`?

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+        blank=True,
+        on_delete=models.SET_NULL,
+    )
+    hosts_naming_template = models.CharField(
```
add TODO to convert it to use `AssetLastHostname` here in the future

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+            'h201.dc ... h299.dc h400.dc h401.dc'
+        ),
+    )
+    next_server = models.CharField(
```
maybe `dhcp_next_server`?

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+        default='',
+        help_text=_('The address for a TFTP server for DHCP.'),
+    )
+    domain = models.CharField(
```
what is this? where is this used?

ar4s replied:
only for information (see prod)

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+        blank=True,
+        default='',
+    )
+    terminators = models.ManyToManyField(
```
afair we should link it with `BaseObject`, not separated model with char field. look at `BaseObjectForeignKey`

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+    )
+    dhcp_broadcast = models.BooleanField(
+        verbose_name=_('Broadcast in DHCP configuration'),
+        default=False,
```
why not true?

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+    )
+    min_ip = models.DecimalField(
+        verbose_name=_('smallest IP number'),
+        null=True,
```
if address is not null, how could this be null?

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+        return network
+
+
+class Network(
```
i'd property `data_center` here too (through `network_environment`)

ar4s replied:
done

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+
+    def search_networks(self):
+        """
+        Search networks (ancestors) order by smallest.
```
order by smallest -> order first by min_ip descending, then by max_ip ascending, to get smallest ancestor netowork containing current network

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+        help_text=_('Presented as string.'),
+        unique=True,
+        blank=False,
+        null=True,
```
`null=True`?

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+    hostname = models.CharField(
+        verbose_name=_('Hostname'),
+        max_length=255,
+        null=True,
```
if it should be nullable, you should use `NullableCharField`

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,416 @@
+
+    def reserve_margin_addresses(self, bottom_count=None, top_count=None):
+        ips = []
+        exists_ips = set(IPAddress.objects.filter(
```
existing_ips

mkurek commented on src/ralph/networks/fields.py:
```diff
@@ -0,0 +1,36 @@
+        raise ValidationError(exc.message)
+
+
+class IPNetwork(Field):
```
add some comment why is this field necessary (ipv6 right?)

ar4s replied:
done

mkurek commented on src/ralph/networks/fields.py:
```diff
@@ -0,0 +1,38 @@
+        super().__init__(*args, **kwargs)
+
+    def db_type(self, connection):
+        return CharField(max_length=50).db_type(connection)
```
why 50?

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,414 @@
+
+    def _assign_ips_to_network(self):
+        IPAddress.objects.exclude(
+            network=self
```
add condition for descendants here, sth like `exclude(network__in=self.get_subnetworks())` - we don't want to reassign all IPs in big network (ex. from subnetworks)

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,414 @@
+        return len(to_create), existing_ips - to_create
+
+    def get_first_free_ip(self):
+        used_ips = list(
```
list -> set, no need to order if we only check if sth is in this set

mkurek commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,414 @@
+        max_ip = int(self.max_ip + 1 if self.netmask == 31 else self.max_ip)
+        ip_as_int = None
+        for ip_as_int in range(min_ip, max_ip):
+            if ip_as_int not in used_ips:
```
shouldn't we check here if ip is reserved too?

ar4s committed:
Added network module

    Network module contains models for network and IP and basic IP Address
    Management.

mkurek commented on src/ralph/data_center/migrations/0010_auto_20160316_1150.py:
```diff
@@ -0,0 +1,58 @@
+                address=asset.management_ip,
+            )
+            ip.number = int(ipaddress.ip_address(ip.address))
+        ip.hostname = asset.management_hostname
```
shouldn't you check here `is_management` flag?

mkurek commented on src/ralph/data_center/migrations/0010_auto_20160316_1150.py:
```diff
@@ -0,0 +1,58 @@
+    IPAddress = apps.get_model('networks', 'IPAddress')
+    ips = IPAddress.objects.exclude(is_management=False, base_object=None)
+    for ip in ips:
+        if not ip.base_object:
```
no need to check that (you've already excluded such objects)

mkurek commented:
:ok:

xliiv commented on src/ralph/data_importer/resources.py:
```diff
@@ -213,6 +215,14 @@ class IPAddressResource(RalphModelResource):
     class Meta:
         model = networks.IPAddress
 
+    def skip_row(self, instance, original):
```
What's the purpose of this?

mkurek replied:
this is built-in method of django-import-export. The purpose of it (in our case) is to skip importing IPAddress if it already exists in database, to, for example, not re-assign it's base object. But - when i'm thinking about it right now, there should be additional logic here - either we should check if `id` column for this ip is filled in csv (purpose: update record, we should pass it) or if this is import from ralph 2.

xliiv replied:
+1 to posibility of updating

xliiv commented on src/ralph/networks/models/networks.py:
```diff
@@ -0,0 +1,426 @@
+        Search networks (ancestors) order first by min_ip descending,
+        then by max_ip ascending, to get smallest ancestor network
+        containing current network.
+        """
```
:+1:

ghost commented:
:ok: I think "update" feature during import can be done in seperate pull request.

szok commented:
:ok:
